// Copyright 2018, Beeri 15.  All rights reserved.
// Author: Roman Gershman (romange@gmail.com)
//
syntax = "proto2";

import "google/protobuf/descriptor.proto";

package tutorial;

option cc_generic_services = true;
option java_package = "com.example.tutorial";
option java_outer_classname = "AddressBookProtos";
option cc_enable_arenas = true;

extend google.protobuf.FieldOptions {
  optional string fd_type = 10101;
  optional string fd_name = 10102;
}

message Address {
  optional string street = 1;
}

message BankAccount {
  repeated int32 activity_id = 3;
  optional string bank_name = 1;
  optional Address address = 2;
}

message Person {
  required string name = 1;
  required int64 id = 2;        // Unique ID number for this person.
  optional string email = 3;

  enum PhoneType {
    MOBILE = 0;
    HOME = 1;
    WORK = 2;
  }

  message PhoneNumber {
    required string number = 1;
    optional PhoneType type = 2 [default = HOME];
  }

  repeated PhoneNumber phone = 4;
  repeated PhoneNumber phone2 = 5;
  optional BankAccount account = 6;
  repeated string tag = 7;
  required double dval = 8;
  optional float fval = 9;
}

// Our address book file is just one of these.
message AddressBook {
  repeated Person person = 1;
  repeated int64 ts = 2;
  repeated uint64 tmp = 3;
  map<int32, string> ids = 4;
  optional int64 fd1 = 5  [ (fd_name) = "another_name"];
}

message JsonParse {
  optional bool bval = 1;
}

service TestService {
  rpc Test(Person) returns (AddressBook);
}
